### PR TITLE
feat: support ed25519 masternode keys 

### PR DESCRIPTION
**Pull request by HashEngineering**
## Issue being fixed or feature implemented

This PR adds support for Ed25519 keys.  These keys are based on a different elliptical curve -- the Twisted Edwards 25519 curve.  This twisted curve is not the same kind of curve used by ECKey.  At first, I thought I could reuse the code of ECKey, but replace the Curve parameters with "curve25519".  When I did this, the public keys didn't match the unit tests I used.  Then it became known that curve25519 and Ed25519 are not the same, though apparently, it is possible to convert between the two.

For the `Ed25519Key` class, we will use the `Ed25519*` classes in Bouncy Castle.  The Bouncy Castle classes take 32 byte private keys and 32 byte public keys.  However, according to SLIP-0011, Ed25519 will be using 33 byte public keys, which have a 0x00 byte prepended to the public key.  This will align the public key size with ECKey.

SLIP-0010 also defines how keys are derived from the passphrase or mnemonic seed.  Another difference with Ed25519 is that SLIP-0010 only explains hardened key derivation.  Child keys are always derived from the private key, never from the public key.  This must be related to a property of the secp256k1 elliptic curve that is not present with the Ed25519 twisted curve.  With secp256k1, public keys can be derived from public keys.

## Checklist:

- [ ] I have performed a self-review of my own code
- [ ] I have commented my code, particularly in hard-to-understand areas
- [ ] I have added or updated relevant unit/integration/functional/e2e tests
- [ ] I have made corresponding changes to the documentation

**For repository code-owners and collaborators only**
- [ ] I have assigned this pull request to a milestone

### CODE REVIEW

File: core/src/main/java/org/bitcoinj/crypto/KeyType.java [hunk 1]
```diff
@@ -20,7 +20,7 @@ public enum KeyType {
     RSA(0),
     ECDSA(1), // secp256k1
     BLS(2),   // BLS12-381
-    EDDSA(3); // ED25519
+    EdDSA(3); // ED25519
```
Comment (HashEngineering) [inline core/src/main/java/org/bitcoinj/crypto/KeyType.java:23]: The `Ed` stands for Edwards (one word) where as EC stands for Elliptic Curve (two words).  

Bouncy Castle also uses `Ed` as the start of all related classes.

File: core/src/main/java/org/bitcoinj/crypto/bls/BLSDeterministicKey.java [hunk 2]
```diff
@@ -94,12 +94,10 @@ public BLSDeterministicKey(ImmutableList<ChildNumber> childNumberPath,
             bytes = serialize(null, false, Script.ScriptType.P2PKH);
             extendedPrivateKey = ExtendedPrivateKey.fromBytes(bytes);
             extendedPublicKey = extendedPrivateKey.getExtendedPublicKey();
-            this.parentFingerprint = (parent != null) ? (int)extendedPrivateKey.getParentFingerprint() : 0;
         } else {
             bytes = serialize(null, true, Script.ScriptType.P2PKH);
             extendedPrivateKey = null;
             extendedPublicKey = ExtendedPublicKey.fromBytes(bytes);
-            this.parentFingerprint = (parent != null) ? (int)extendedPublicKey.getParentFingerprint() : 0;
```
Comment (HashEngineering) [inline core/src/main/java/org/bitcoinj/crypto/bls/BLSDeterministicKey.java:102]: This resolved an outstanding issue from the previous PR: https://github.com/dashpay/dashj/pull/199#discussion_r1133025331

These lines are removed as `parentFingerprint` is assigned in line 91.

File: core/src/main/java/org/bitcoinj/wallet/AnyDeterministicKeyChain.java [hunk 3]
```diff
@@ -233,6 +237,7 @@ private int calcDefaultLookaheadThreshold() {
         protected IDeterministicKey spendingKey = null;
         protected ImmutableList<ChildNumber> accountPath = null;
         protected KeyFactory keyFactory = ECKeyFactory.get();
+        protected boolean hardenedKeysOnly = false;
```
Comment (HashEngineering) [inline core/src/main/java/org/bitcoinj/wallet/AnyDeterministicKeyChain.java:240]: For some purposes, we only want hardened keys or we can only create hardened keys (Ed25519 only supports hardened key derivation according to SLIP-0010).

File: core/src/test/java/org/bitcoinj/wallet/authentication/AuthenticationKeyChainGroupTest.java [hunk 4]
```diff
@@ -166,8 +188,10 @@ public void serializationTest() throws UnreadableWalletException {
         assertEquals(wallet.currentAuthenticationKey(AuthenticationKeyChain.KeyChainType.MASTERNODE_OWNER),
                 walletCopy.currentAuthenticationKey(AuthenticationKeyChain.KeyChainType.MASTERNODE_OWNER));
 
-        // TODO: for some reason "wallet" has creation times for all BLS keys, so this fails if we compare keys, so lets compare priv key bytes
```
Comment (HashEngineering) [inline core/src/test/java/org/bitcoinj/wallet/authentication/AuthenticationKeyChainGroupTest.java:169]: This bug was fixed in a previous PR, but the comment remained.

File: core/src/main/java/org/bitcoinj/wallet/WalletProtobufSerializer.java [hunk 5]
```diff
@@ -281,6 +282,13 @@ public Protos.Wallet walletToProto(Wallet wallet) {
             extendedKeyChainBuilder.addAllKey(wallet.getProviderOperatorKeyChain().serializeToProtobuf());
             extendedKeyChains.add(extendedKeyChainBuilder.build());
         }
+        if(wallet.getProviderPlatformOperatorKeyChain() != null) {
+            Protos.ExtendedKeyChain.Builder extendedKeyChainBuilder = Protos.ExtendedKeyChain.newBuilder();
+            extendedKeyChainBuilder.setKeyType(Protos.ExtendedKeyChain.KeyType.EDDSA);
+            extendedKeyChainBuilder.setType(Protos.ExtendedKeyChain.ExtendedKeyChainType.MASTERNODE_PLATFORM_OPERATOR);
+            extendedKeyChainBuilder.addAllKey(wallet.getProviderPlatformOperatorKeyChain().serializeToProtobuf());
+            extendedKeyChains.add(extendedKeyChainBuilder.build());
+        }
```
Comment (HashEngineering) [inline core/src/main/java/org/bitcoinj/wallet/WalletProtobufSerializer.java:291]: This is a "manual" addition to support the new key type.  

In a separate PR, this serialization will be refactored into the `AuthenticationGroupExtension` class.

File: core/src/main/java/org/bitcoinj/wallet/AuthenticationKeyChain.java [hunk 6]
```diff
@@ -380,10 +413,6 @@ public KeyChainType getType() {
         return type;
     }
 
-    public void setHardenedChildren(boolean hardenedChildren) {
-        this.hardenedChildren = hardenedChildren;
-    }
```
Comment (HashEngineering) [inline core/src/main/java/org/bitcoinj/wallet/AuthenticationKeyChain.java:385]: this property is being moved to a super class.

**Review by Syn-McJ — APPROVED** [review]
Looks great

**Pull request merged**